File: Sources/HerdrMobile/Console/ConsoleAgent.swift
import Foundation

/// One row of the Console (#8): an Agent joined with its Host identity and
/// workspace context. The list is flat across Hosts; the workspace is a
/// context tag only, never a grouping level.
struct ConsoleAgent: Identifiable, Sendable, Equatable {
    /// Pane addresses are unique per herdr session, not across Hosts; the
    /// row identity pairs them.
    struct ID: Hashable, Sendable {
        let hostID: Host.ID
        let paneID: String
    }

    let hostID: Host.ID
    let hostName: String
    var agent: Agent
    /// Workspace label from the session snapshot; nil when the snapshot did
    /// not carry the workspace.
    let workspaceLabel: String?
    /// Worktree repo name where the workspace has one — sharper context than
    /// a raw checkout path.
    let repoName: String?
    /// Trailing terminal output (`pane.read`, ANSI stripped), fetched after
    /// snapshots and status changes; nil until the first read lands.
    var lastOutputSnippet: String?

    var id: ID { ID(hostID: hostID, paneID: agent.paneID) }
}

extension AgentStatus {
    /// Console sort bucket: Blocked > Working > everything else. Idle, Done,
    /// Unknown, and any status this build does not recognize (herdr's API
    /// has no stability guarantee) share the bottom bucket — a status we
    /// cannot interpret is not actionable, so it must not outrank one we can.
    var consoleSortBucket: Int {
        switch self {
        case .blocked: 0
        case .working: 1
        default: 2
        }
    }
}

extension [ConsoleAgent] {
    /// The Console order: status bucket first, then stable Host/workspace/
    /// pane keys so rows never jitter between equal statuses.
    func consoleSorted() -> [ConsoleAgent] {
        sorted { lhs, rhs in
            let lhsBucket = lhs.agent.status.consoleSortBucket
            let rhsBucket = rhs.agent.status.consoleSortBucket
            if lhsBucket != rhsBucket { return lhsBucket < rhsBucket }
            if lhs.hostName != rhs.hostName { return lhs.hostName < rhs.hostName }
            if lhs.hostID != rhs.hostID {
                return lhs.hostID.uuidString < rhs.hostID.uuidString
            }
            let lhsWorkspace = lhs.workspaceLabel ?? ""
            let rhsWorkspace = rhs.workspaceLabel ?? ""
            if lhsWorkspace != rhsWorkspace { return lhsWorkspace < rhsWorkspace }
            return lhs.agent.paneID < rhs.agent.paneID
        }
    }
}

File: android/app/src/main/java/dev/bybee/heeler/console/ConsoleAgent.kt
package dev.bybee.heeler.console

import dev.bybee.heeler.core.transport.Agent
import dev.bybee.heeler.core.wire.AgentStatus
import java.util.Locale

/** A Host-qualified Pane address. Pane IDs are only unique within one Host. */
data class ConsoleAgentId(
    val hostId: String,
    val paneId: String,
)

/**
 * One flat Console row: an Agent joined to its Host and workspace context.
 * Workspace context decorates a row; it never introduces a second grouping.
 */
data class ConsoleAgent(
    val hostId: String,
    val hostName: String,
    val paneId: String,
    val terminalId: String,
    val kind: String,
    val name: String?,
    val title: String,
    val status: AgentStatus,
    val workspaceId: String,
    val workspaceLabel: String?,
    val repoName: String?,
    val checkoutPath: String?,
    val cwd: String,
    val revision: Int,
    val lastOutputSnippet: String? = null,
    /** Connection identity used to reject work returned by an older session. */
    val connectionGeneration: Long,
) {
    val id: ConsoleAgentId get() = ConsoleAgentId(hostId, paneId)
    val displayName: String get() = name?.takeIf(String::isNotBlank) ?: kind
    val switcherLabel: String get() = workspaceLabel ?: repoName ?: displayName
    val skillsProjectRoot: String? get() = checkoutPath?.takeIf(String::isNotBlank)
        ?: cwd.takeIf(String::isNotBlank)

    companion object {
        fun fromAgent(
            hostId: String,
            hostName: String,
            agent: Agent,
            workspaceLabel: String?,
            repoName: String?,
            checkoutPath: String?,
            lastOutputSnippet: String?,
            connectionGeneration: Long,
        ): ConsoleAgent = ConsoleAgent(
            hostId = hostId,
            hostName = hostName,
            paneId = agent.paneID,
            terminalId = agent.terminalID,
            kind = agent.kind,
            name = agent.name,
            title = agent.title,
            status = agent.status,
            workspaceId = agent.workspaceID,
            workspaceLabel = workspaceLabel,
            repoName = repoName,
            checkoutPath = checkoutPath,
            cwd = agent.cwd,
            revision = agent.revision,
            lastOutputSnippet = lastOutputSnippet,
            connectionGeneration = connectionGeneration,
        )
    }
}

/** Workspace information retained from a Host's latest successful snapshot. */
data class ConsoleWorkspace(
    val id: String,
    val label: String,
)

/** Blocked, Done, Working, Idle, then unknown server values. */
val AgentStatus.consoleSortBucket: Int
    get() = when (rawValue) {
        AgentStatus.blocked.rawValue -> 0
        AgentStatus.done.rawValue -> 1
        AgentStatus.working.rawValue -> 2
        AgentStatus.idle.rawValue -> 3
        else -> 4
    }

/** Stable Console ordering prevents equal-status cards from visually jittering. */
fun List<ConsoleAgent>.consoleSorted(): List<ConsoleAgent> = sortedWith(
    compareBy<ConsoleAgent> { it.status.consoleSortBucket }
        .thenBy { it.hostName.lowercase(Locale.ROOT) }
        .thenBy(ConsoleAgent::hostId)
        .thenBy { it.workspaceLabel.orEmpty() }
        .thenBy(ConsoleAgent::paneId),
)
